source/common/router/router_ratelimit.cc

```diff
@@ -44,27 +67,29 @@ bool DynamicMetadataRateLimitOverride::populateOverride(
 }
 
 bool SourceClusterAction::populateDescriptor(const Router::RouteEntry&,
-                                             RateLimit::Descriptor& descriptor,
+                                             RateLimit::DescriptorEntry& descriptor_entry,
                                              const std::string& local_service_cluster,
                                              const Http::HeaderMap&,
                                              const Network::Address::Instance&,
                                              const envoy::config::core::v3::Metadata*) const {
-  descriptor.entries_.push_back({"source_cluster", local_service_cluster});
+  descriptor_entry.key_ = "source_cluster";
+  descriptor_entry.value_ = local_service_cluster;
```

Comment:
nit: I would just add a constructor to `DescriptorEntry` then you can revert all of these changes.

Reply:
Yeah, I'm not sure why this refactoring was needed...

Reply:
To be able to support PopulateLocalDescriptor function. LocalDescriptor just have DescriptorEntry and not the override...

Reply:
We could just ignore nilopt override, no?

Reply:
I think override also makes sense for local rate limiting, although it's not exposed as such.

Reply:
> We could just ignore nilopt override, no?
We would have to use Descriptor and then convert Descriptor to LocalDescriptor in that case...

> I think override also makes sense for local rate limiting, although it's not exposed as such.
yeah we added tokenbucket for LocalDescriptor to keep it in sync with local rate limiting design...

Reply:
Maybe, I'm confused, but I think descriptors don't really care whether they are local or global. The difference is where the bucket is calculated. But the code to populate descriptors should be the same.

Reply:
If I understood currently, you are thinking what I had implemented earlier!
RateLimitDescriptor be part of local rate limit proto, rather than creating a new LocalRateLimitDescriptor.
But then it had override stuff too and thus it was refactored to have a new LocalRateLimitDescriptor which contains DescriptorEntry and token bucket.
ratelimit.h file has structs corresponding to the above proto messages...

Now, DescriptorEntry doesn't care if it's local or global.
For global there is DescriptorEntry and Override
For local there is DescriptorEntry and TokenBucket.
Thus, code to populate DescriptorEntry is shared between local and global...

hope the above description helps..

Reply:
I think the current implementation is mostly what we want. LocalDescriptor is just a version of descriptor which doesn't include extraneous things that we don't need. All I'm asking for here is to reduce changes that are not needed via adding a constructor.

Reply:
That's fine. I anticipate merge conflict with another rate limiting PR, so I'll hold on to reconciling the headers.

Reply:
LocalDescriptor and Descriptor are not the same thing though. Do you want to templatize this? Or just pass a vector reference to entries?

Reply:
Sorry I was mainly thinking allowing something like `descriptor_entry = {...}`, but it's really just a nit. Feel free to ignore if you don't think it's worth it.

Reply:
Done.